load-test-framework/src/main/java/com/google/pubsub/flic/controllers/Client.java

```diff
@@ -0,0 +1,239 @@
+        return "gcloud";
+      case KAFKA_PUBLISHER:
+      case KAFKA_SUBSCRIBER:
+        return "kafka";
+    }
+    return null;
+  }
+
+  ClientType getClientType() {
+    return clientType;
+  }
+
+  ListenableFuture<Void> getDoneFuture() {
+    return doneFuture;
+  }
+
+  private LoadtestGrpc.LoadtestStub getStub() {
+    return LoadtestGrpc.newStub(
+        ManagedChannelBuilder.forAddress(networkAddress, port).usePlaintext(true).build());
+  }
+
+  synchronized long[] getBucketValues() {
+    return bucketValues;
+  }
+
+  void start() throws Throwable {
+    // Send a gRPC call to start the server
+    log.info("Connecting to " + networkAddress + ":" + port);
+    StartRequest.Builder requestBuilder = StartRequest.newBuilder()
+        .setProject(project)
+        .setTopic(topic)
+        .setBatchSize(batchSize)
+        .setMaxOutstandingRequests(maxOutstandingRequests)
+        .setMessageSize(messageSize)
+        .setRequestRate(requestRate)
+        .setStartTime(startTime);
+    if (numberOfMessages > 0) {
+      requestBuilder.setNumberOfMessages(numberOfMessages);
+    } else {
+      requestBuilder.setStopTime(Timestamp.newBuilder()
+          .setSeconds(Math.max(startTime.getSeconds(), System.currentTimeMillis() / 1000) +
+              loadtestLengthSeconds).build());
+    }
+    switch (clientType) {
+      case CPS_GCLOUD_SUBSCRIBER:
+        requestBuilder.setPubsubOptions(PubsubOptions.newBuilder()
+            .setMaxMessagesPerPull(10)
+            .setSubscription(subscription));
+        break;
+      case KAFKA_PUBLISHER:
+      case KAFKA_SUBSCRIBER:
+        requestBuilder.setKafkaOptions(KafkaOptions.newBuilder()
+            .setBroker(broker)
+            .setPollLength(100));
+        break;
+    }
+    StartRequest request = requestBuilder.build();
+    SettableFuture<Void> startFuture = SettableFuture.create();
+    stub = getStub();
+    stub.start(request, new StreamObserver<StartResponse>() {
+      private int connectionErrors = 0;
+      @Override
+      public void onNext(StartResponse response) {
+        log.info("Successfully started client [" + networkAddress + "]");
+        clientStatus = ClientStatus.RUNNING;
+      }
+
+      @Override
+      public void onError(Throwable throwable) {
+        if (connectionErrors > 10) {
+          log.error("Client failed to start " + connectionErrors + " times, shutting down.");
+          clientStatus = ClientStatus.FAILED;
+          startFuture.setException(throwable);
+          doneFuture.setException(throwable);
+          return;
+        }
+        connectionErrors++;
+        try {
+          Thread.sleep(5000);
+        } catch (InterruptedException e) {
+          log.info("Interrupted during back off, retrying.");
+        }
+        log.debug("Going to retry client connection, likely due to start up time.");
+        stub = getStub();
+        stub.start(request, this);
+      }
+
+      @Override
+      public void onCompleted() {
```

Comment:
What's the difference between onNext() and onCompleted() for non-streaming? It seems like both startFuture.set(null) and clientStatus = ClientStatus.RUNNING should be done in the same callback.